### PR TITLE
db/system_keyspace: table with connected cql clients

### LINKED ISSUE #4820: Provide information about clients connected
While we have counters for the number of open connections it would help to extract the info on the clients connected and as much info as possible (protocol version / ssl / user / driver / driver versionn / cql version etc.)

https://issues.apache.org/jira/browse/CASSANDRA-13665

yet extend it to extract all the info from auth state and also from the STARTUP phase that does include additional info 

Screenshot from 2019-08-09 00-25-25

It would be best to expose this as a system table and not only as a nodetool command

### PR DESCRIPTION
**Pull request by jul-stas**
This change introduces system.clients table, which provides information about CQL clients connected.

PK is the client's IP address, CK consists of outgoing port number and client_type (which will be extended in future to thrift/alternator/redis). Table supplies also shard_id and username. Other columns, like connection_stage, driver_name, driver_version..., are currently empty but exist for C* compatibility and future use.

This is an ordinary table (i.e. non-virtual) and it's updated upon accepting connections. This is also why C*'s column request_count was not introduced. In case of abrupt DB stop, the table should not persist, so it's being truncated on startup.
Resolves #4820

### CODE REVIEW

File: connection_notifier.cc [hunk 1]
```diff
@@ -55,9 +55,3 @@ future<> notify_disconnected_client(gms::inet_address addr, client_type ct, int
                      db::system_keyspace::clients_cf_name());
     return db::execute_cql(req, addr.addr(), port, to_string(ct)).discard_result();
 }
-
```
Comment (haaawk) [inline connection_notifier.cc]: This shouldn't be here at all. You're adding the function in previous patch and then removing it here. Just don't add the method at all please.

File: connection_notifier.hh [hunk 2]
```diff
@@ -49,5 +49,3 @@ struct client_data {
 
 future<> notify_new_client(client_data cd);
 future<> notify_disconnected_client(gms::inet_address addr, client_type ct, int port);
-
-future<> clear_clientlist();
```
Comment (haaawk) [inline connection_notifier.hh]: Don't add this at all please.

File: main.cc [hunk 3]
```diff
@@ -1056,6 +1056,13 @@ int main(int ac, char** av) {
                 view_builder.invoke_on_all(&db::view::view_builder::start).get();
             }
 
+            // Truncate `clients' CF - this table should not persist between server restarts.
```
Comment (haaawk) [inline main.cc]: Please wrap it into a function with a self describing name to hide the complexity in this already quite long function.

File: service/storage_service.hh [hunk 4]
```diff
@@ -213,6 +213,10 @@ public:
         return _db;
     }
 
+    shared_ptr<distributed<cql_transport::cql_server>> get_dist_cql_server() const {
```
Comment (haaawk) [inline service/storage_service.hh]: Where is this used? I couldn't find the place.

**Comment by haaawk** [comment]
LGTM in general. One question though, have you run all unit tests on this change? You're adding a new table to system keyspace and this should probably cause schema_change_test to fail so I'm a bit surprise it does not.

**Comment by jul-stas** [comment]
What's going on here:
As noted by Rafael, `batch_test.py:TestBatch.replay_after_schema_change_test` used to fail after merging #5233. This is my early attempt to fix it. Only difference is the way `system.clients` gets truncated: instead of:
```
-    const static sstring req
-            = format("TRUNCATE system.{};", db::system_keyspace::clients_cf_name());
-    return db::execute_cql(req).discard_result();
```
...now we have this:
```
+            proxy.local().get_db().local().truncate(
+                    proxy.local().get_db().local().find_keyspace("system"),
+                    proxy.local().get_db().local().find_column_family("system", "clients"), [] {
+                return make_ready_future<db_clock::time_point>(db_clock::now());
+            }, false /* with_snapshot */).get();
```
...which does not propagate the call between endpoints, esp. while system is not fully booted up. Technical changes are only about removing unneeded functions and #includes.

File: service/storage_service.hh (hunk 4 above)
Comment (jul-stas) [reply service/storage_service.hh]: aaah, a ghost from virtual tables. Removed

File: connection_notifier.hh (hunk 2 above)
Comment (jul-stas) [reply connection_notifier.hh]: Ultimately the code from `main()` went here, so I kept it

**Comment by jul-stas** [comment]
> LGTM in general. One question though, have you run all unit tests on this change? You're adding a new table to system keyspace and this should probably cause schema_change_test to fail so I'm a bit surprise it does not.

You already know, but I'll leave the note for other reviewers:
`schema_tables.cc:629 @calculate_schema_digest()` has explicit check that excludes CFs from `system` KS. So tests pass as an expected behavior - a nice surprise.

**Comment by jul-stas** [comment]
@avikivity @nyh Sorry for nagging, I know it's nothing fancy here, but this PR hangs for very long already. Perhaps we could move on with it?

File: transport/server.cc [hunk 5]
```diff
@@ -239,14 +239,21 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
...
-                } catch (...) {
-                    clogger.debug("connection error: {}", std::current_exception());
+                } catch(...) {
+                    clogger.info("exception while advertising new connection: {}", std::current_exception());
                 }
+                return conn->process().finally([this, conn] {
+                    --_connections;
+                    return unadvertise_connection(conn);
+                }).handle_exception([] (std::exception_ptr ep) {
+                    clogger.info("exception found after connection close: {}", ep);
```
Comment (nyh) [inline transport/server.cc]: Is "after connection close" the correct message for something that comes from conn->process()? Why "after connection close"? Note that before your patch the log message was about "connection error".
Also why did you change it from a debug message to an info? (maybe it makes sense, but I don't know why it was debug before).

File: connection_notifier.cc [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+#include "database.hh"
+#include "service/storage_proxy.hh"
+
+#include <stdexcept>
+
+namespace db::system_keyspace {
+const char *const clients_cf_name();
+}
+
+static sstring to_string(client_type ct) {
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.cc]: Sure, there are pros and cons. My main reasoning was that enum is a limited set (harder to misspell) and its values come from single "origin" (so people can look into code and immediately tell what the allowed values are). And saves a few bytes :)
But I won't defend it - if you're convinced about making `ct` a string, please let me know and I'll change it.

File: connection_notifier.hh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    thrift,
+    alternator,
+};
+
+// Representation of a row in `system.clients'. std::optionals are for nullable cells.
+struct client_data {
+    gms::inet_address ip;
+    int32_t port;
+    client_type ct;
+    int32_t shard_id;
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.hh]: Yes, it's the serverside shard_id that handles particular connection. Would you recommend a comment in code or maybe... IDK... to submit a full-blown table description to docs?

File: connection_notifier.cc [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+#include "connection_notifier.hh"
+#include "db/query_context.hh"
+#include "cql3/constants.hh"
+#include "database.hh"
+#include "service/storage_proxy.hh"
+
+#include <stdexcept>
+
+namespace db::system_keyspace {
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.cc]: I'm sorry, I don't understand. You mean I should put the decl of `clients_cf_name()` into `system_keyspace.hh`? Or create a brand new header for one function?
BTW `system_keyspace::CLIENTS` is, like all table names, `constexpr` and I don't know of a way to forward-declare a constexpr variable (might be impossible?)

File: transport/server.cc [hunk 9]
```diff
@@ -239,14 +239,21 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
             auto conn = make_shared<connection>(*this, server_addr, std::move(fd), std::move(addr));
             ++_connects;
             ++_connections;
-            // Move connection to the background, monitor for lifetime and errors.
-            (void)conn->process().then_wrapped([this, conn] (future<> f) {
-                --_connections;
+            // Move conn to the background, announce, then block while monitoring for lifetime/errors.
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply transport/server.cc]: Right, but... to me, personally, `log` suggests just printing and not much else...
How about `register_...` or maybe `new_conn_notify_clients_table`?

File: connection_notifier.cc [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        default: throw std::runtime_error("Invalid client_type");
+    }
+}
+
+future<> notify_new_client(client_data cd) {
+    const static sstring req
+            = format("INSERT INTO system.{} (address, port, client_type, shard_id, protocol_version, username) "
+                     "VALUES (?, ?, ?, ?, ?, ?);", db::system_keyspace::clients_cf_name());
+    
+    return db::execute_cql(req,
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.cc]: Can we do that in a follow-up?
I would hesitate to prepare those statements without a benchmark. Not to mention that... this whole PR is temporary solution (until somebody understands/refactors virtual readers), so maybe it's not worth it..?

File: connection_notifier.hh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    std::optional<sstring> connection_stage;
```
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.hh]: Yes, I planned to fill them in future, so I left "placeholders". Added an explanatory comment.

File: transport/server.cc (hunk 5 above)
Comment (jul-stas) [reply transport/server.cc]: Ouch, absolutely, "close" is not the only thing that happens here. I updated exception description.
Also, I bumped loglevel because I think that connection errors are a severe thing and should not be buried in debug messages, as it used to be.

**Comment by slivne** [comment]
@jul-stas did you address everything - if you have we can ask for a re-review

**Comment by jul-stas** [comment]
@slivne It depends on what Nadav @nyh will respond to those two:
https://github.com/scylladb/scylla/pull/5351#discussion_r352594340
https://github.com/scylladb/scylla/pull/5351#discussion_r353337700
Apart from that - yes.

File: connection_notifier.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (nyh) [reply connection_notifier.cc]: Maybe I'm missing something but I don't understand why this function is needed at all given that you have the variable `system_keyspace::CLIENTS`. Or if it's needed (?), why it's in conneciton_notifier.cc and not some inline function in the header file.

File: connection_notifier.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (nyh) [reply connection_notifier.cc]: I don't know. It seems overengineered to me, but I guess it's also a matter of style. So I'll withdraw my comments.

File: connection_notifier.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (nyh) [reply connection_notifier.cc]: This patch, temporary or not, will make connection establishment slower as each connection is logged. I *think* that because most clients will make an effort to only create a few connections, this overhead will be negligible, so your code is fine - but I don't really know.
I think it's important that if you feel that prepared statements could be better here but just don't want to do it now - that you should add a FIXME comment in the code, otherwise we'll never remember to go to it.

File: connection_notifier.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.cc]: You're right, I removed this function in favor of extern variable.
The only downside is a lost `constexpr` qualifier, so no big deal, I think.

File: connection_notifier.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (jul-stas) [reply connection_notifier.cc]: Thanks, I added FIXME for now.
Anyway, I'm expected to write some test here later, so that will be a good opportunity for benchmark/prepare.

**Comment by jul-stas** [comment]
Hi @nyh, I know this how annoying this PR is getting, but ping

**Comment by slivne** [comment]
@nyh ping

**Comment by jul-stas** [comment]
@nyh ping

File: transport/server.cc [hunk 12]
```diff
@@ -237,14 +238,22 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
...
-            // Move connection to the background, monitor for lifetime and errors.
-            (void)conn->process().then_wrapped([this, conn] (future<> f) {
-                --_connections;
+            // Move the processing into the background.
+            (void)futurize_apply([this, conn] {
+                return advertise_new_connection(conn); // Notify any listeners about new connection.
+            }).then_wrapped([this, conn] (future<> f) {
                 try {
                     f.get();
-                } catch (...) {
```
Comment (psarna) [inline transport/server.cc]: please restore the whitespace between `catch` and `(`

Comment (jul-stas) [reply transport/server.cc]: done

**Comment by psarna** [comment]
I'm queueing to next.
